src/zeroband/training/weights.py

```diff
@@ -1,82 +1,146 @@
+        # Rename temporary file to indicate checkpoint is complete
+        tmp_model_path.rename(model_path)
+
+        # Save model config, generation arguments and tokenizer
+        model.config.save_pretrained(step_path)
+        model.generation_config.save_pretrained(step_path)
+        tokenizer.save_pretrained(step_path)
+
+        self._logger.debug(f"Saved weight checkpoint to {step_path} in {time.time() - start_time:.2f} seconds")
+
+    def _save(self, model: Model, tokenizer: AutoTokenizer, dtype: torch.dtype, step: int):
+        """Synchronous helper of `save`."""
+        cpu_state = self._gather_weights(model, dtype)
+        self._save_to_path(cpu_state, model, tokenizer, step)
+
+    def save(
+        self,
+        model: Model,
+        tokenizer: AutoTokenizer,
+        step: int,
+        dtype: torch.dtype = torch.bfloat16,
+    ):
+        """Save a HF-compatible weight-only checkpoint for a given step."""
+
+        if self.config.save_async:
+            thread = threading.Thread(
+                target=self._save,
+                args=(model, tokenizer, dtype, step),
+                name=f"weight-checkpoint-save-{step}",
+            )
+            thread.start()
+        else:
+            self._save(model, tokenizer, dtype, step)
+        return self._get_model_path(step)
+
+    def _clean(self, step: int):
+        """Synchronous helper of `clean`."""
+        candidate_path_to_delete = self._get_step_path(step)
+        self._logger.debug(f"Considering to delete weight checkpoint {candidate_path_to_delete}")
+        keep_for_eval = self.config.interval and step % self.config.interval == 0
+        # For checkpointing step x, we need all weight checkpoints in [x-async_level, x] (for logprob model)
+        # To get [n-k, n] with interval n and buffer k over all natural numbers x, we use the condition (n - (x % n)) % n <= k
+        keep_for_ckpt = (
+            self.ckpt_config
+            and (self.ckpt_config.interval - (step % self.ckpt_config.interval)) % self.ckpt_config.interval
+            <= self.async_level
+        )
+        if not (keep_for_eval or keep_for_ckpt):
+            self._logger.debug(
+                f"Removing past weight checkpoint {candidate_path_to_delete} ({keep_for_eval=}, {keep_for_ckpt=})"
+            )
+            shutil.rmtree(candidate_path_to_delete, ignore_errors=True)
+
+    def clean(self, step: int):
+        """
+        Considers deleting a weight checkpoint for a given step. There are two reasons not to delete a checkpoint:
+        1. The step is an evaluation step (e.g. step % weights.interval == 0)
+        2. The step is a checkpoint step or at most async_level steps earlier
+        """
+        if self.config.save_async:
+            thread = threading.Thread(
+                target=self._clean,
+                args=(step,),
+                name=f"weight-checkpoint-clean-{step}",
+            )
+            thread.start()
```

Comment:
Now that we have a class I would add the thread to a list of thread we could decide later to handle the termination of this threads. for now its fine